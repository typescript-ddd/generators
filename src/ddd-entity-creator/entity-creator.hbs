import { DomainEventBus } from "@typescript-ddd/core";
import { {{pascal entityName}}Repository } from "../../infrastructure/persistence";
import { {{pascal entityName}} } from "../../domain";

/**
 * Represents {{noun entityName}} creator.
 */
export class {{pascal entityName}}Creator {
  /**
   * Initializes a new instance of {{noun entityName}} creator.
   * @param {{lbrace}}{{pascal entityName}}Repository{{rbrace}} repository - The {{natural entityName}} repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: {{pascal entityName}}Repository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Creates a new {{natural entityName}}.
   * @param {{lbrace}}Create{{pascal entityName}}Props{{rbrace}} props - the properties to create the {{natural entityName}}.
   * @returns {{lbrace}}Promise<{{pascal entityName}}>{{rbrace}} - A newly created {{natural entityName}}.
   */
  public async create(props: Create{{pascal entityName}}Props): Promise<{{pascal entityName}}> {
    const entity = {{pascal entityName}}.create(props);
    await this.repository.save(entity);
    this.eventBus.publish(entity.pullEvents());
    return entity;
  }
}
