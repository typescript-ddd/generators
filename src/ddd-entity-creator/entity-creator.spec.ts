import { EntityCreatorTemplateData } from "./entity-creator.types";
import { generateEntityCreator } from "./entity-creator.generator";

describe("EntityCreator", () => {
    it("should render a created entity", () => {
        // Arrange
        const data: EntityCreatorTemplateData = {
        entityName: "User"
        };
    
        // Act
        const result = generateEntityCreator(data);
    
        // Assert
        expect(result).toMatchSnapshot();
    });
    });