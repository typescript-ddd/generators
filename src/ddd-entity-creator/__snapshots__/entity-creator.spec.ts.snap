// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityCreator should render a created entity 1`] = `
"import { DomainEventBus } from "@typescript-ddd/core";
import { UserRepository } from "../../infrastructure/persistence";
import { User } from "../../domain";

/**
 * Represents a user creator.
 */
export class UserCreator {
  /**
   * Initializes a new instance of a user creator.
   * @param {UserRepository} repository - The user repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: UserRepository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Creates a new user.
   * @param {CreateUserProps} props - the properties to create the user.
   * @returns {Promise<User>} - A newly created user.
   */
  public async create(props: CreateUserProps): Promise<User> {
    const entity = User.create(props);
    await this.repository.save(entity);
    this.eventBus.publish(entity.pullEvents());
    return entity;
  }
}
"
`;
